### PR TITLE
IntrusiveHashMap: Refresh TSHashTable for C++ eleventy.

### PR DESCRIPTION
**Pull request by SolidWallOfCode**
This is a replacement for `TSHashTable` and eventually TCL hash tables.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1151,6 +1152,7 @@ add_library(libtsutil SHARED
         lib/ts/Tokenizer.h
         lib/ts/Trie.h
         lib/ts/TsBuffer.h
+	lib/ts/ts_meta.h
```
Comment (maskit) [inline CMakeLists.txt]: What is this for?

Comment (SolidWallOfCode) [reply CMakeLists.txt]: There just a few comments in that header file to explain what it does...

It was part of some work that ended up in the bit bucket. This enabled optional types / functions in the template parameter `H`. The point was to allow overrides of certain deduced types. For example, `H::value_type` would override the type deduced from the link accessor methods but would not be required. Unfortunately the additional implementation effort to make that work was, IMHO, too much work for not enough return - the accessors in `H` can handle that, if needed, more easily. Mostly I forgot this was still here, but I do have some other use cases planned and I might well still use it for this work, so I'd rather leave it in. I suppose I could split that out in to a separate commit.

Comment (maskit) [reply CMakeLists.txt]: I spent my precious time to find places that require this :-)

**Comment by SolidWallOfCode** [comment]
If you're going to gang up on me, I'll pull it out, but @shinrich  will need to re-approve then.

**Comment by SolidWallOfCode** [comment]
[approve ci autest]

**Comment by bryancall** [comment]
Why are we introducing another hash table.  This is not what we talked about doing in Cork.  I thought we were going to look at doing a C++ wrapper around Concurrency Kit?

**Comment by SolidWallOfCode** [comment]
As noted in the initial comment, this is a refresh of `TSHashTable`,  it is not a new hash table. Once this PR is committed I will be replace all instance of TSHashTable with this better implementation. I need features in this update that can't wait on the concurrent hash table work, and further this will be used in places where a concurrent hash table is more than is required. I hope at some point to begin replacing uses of the TCL based hash table with this as well (see #3989 for an example).

**Comment by SolidWallOfCode** [comment]
Note that #4029 is part of the work of this upgrade, replacing `TSHashTable` with `IntrusiveHashMap`.

**Comment by bryancall** [comment]
The comment says it is a replacement for TSHashTable, but looking at the commit is an additional hash table.  I see that there is another PR that has plans to remove IntrusiveHashMap, but is not referenced here.

What features does this add that you need?  I worry that we are going down the path of relying on our own data structures that are immature and not well tested and will result in maintenance and incompatibility between versions.

I looked at the other two PRs that are going to use IntrusiveHashMap and it looks like they will be using them in a concurrent access pattern which mean that this hash table is solving the wrong problem.

**Comment by SolidWallOfCode** [comment]
Some of the features are:

*  Use of C++ 11 for cleaner and more robust code.
*  Shifting the descriptor structure to require fewer pieces and more in line with other uses in TS.
*  Integration with `InstrusiveDList` instead of the `DLL` list, again for cleaner and more consistent code.
*  Remove the obscure and specialized `Location` class and use the STL iterator range style API.

What I really need this for are other uses, e.g. the work to eliminate `tsconfig` and use of TCL hashes. However, since it was agreed that we needed to consolidate hash tables, I did this as a `TSHashTable` replacement that could be use both for that other work and to replace current uses.

For many of the other uses, the intrusive nature of the hash map is critical, as I want to use it in conjunction with `MemArena`. There is therefore a requirement that all memory for the hash map is allocated from the `MemArena`. This class makes that possible.

This isn't needed for most (all?) of the replacement uses, but I have suggested use of classes such as `std::unordered_map` and been strongly rebuffed every single time. If I can't use that, and I have to have this for some specialized uses, I might as well use this class everywhere. I need to also use it to replace `TSHashTable` since I can't add more hash containers.

> I looked at the other two PRs that are going to use IntrusiveHashMap and it looks like they will be using them in a concurrent access pattern which mean that this hash table is solving the wrong problem.

What then should I do? Not replace `TSHashTable` nor the TCL based `ink_hash_table` in those uses? Your original complaint, as far as I can tell, was that I hadn't. Therefore I worked on those PRs and linked them here. Or is it that I shouldn't do any work in this area until someone does the full concurrent hash table work?

**Comment by bryancall** [comment]
We should use `std::unordered_map` instead of using `TSHashTable` and creating another non-concurrent hash table for hostdb and session manager.

I don't see how this has a relation to getting rid of `tsconfig`.

**Comment by SolidWallOfCode** [comment]
[approve ci autest]

**Pull request merged**